BigQuery: Allow a URI as the storage url for a load job

jeremywadsack opened the pull request:
In the [BigQuery `load_job` documentation](http://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-bigquery/v0.30.0/google/cloud/bigquery/table?method=load_job-instance), it states that you can use "a file or the URI of a Google Cloud Storage file", which I understood to be a [`URI`](https://docs.ruby-lang.org/en/2.4.0/URI.html) type object. 

After having this fail, I noticed that the docs do list the types that the object can be and that `URI` isn't listed. But I thought this could be useful to others (and is to us as we're using a `URI`). 

This pull request updates the method to accept a `URI` type parameter. In fact, this alters the method to accept any object that's `#to_s` returns a string representing a GS url.

quartzmo commented:
@jeremywadsack Thank you for submitting this PR. I think this change seems like a good idea. It will also need to be applied to `Table#load`, `Dataset#load_job` and `Dataset#load`, but you might hold off on that until @blowmage has commented.

quartzmo commented:
I'll look into fixing the broken build. It appears to be unrelated to this PR.

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -2010,8 +2014,8 @@ def load_local file, options = {}
           file.respond_to?(:to_gs_url) ||
-            (file.respond_to?(:to_str) &&
-            file.to_str.downcase.start_with?("gs://"))
+            (file.respond_to?(:to_s) &&
```
This is not a good change. The existence of a `to_s` method is not an indication that the object is a string, and nearly all objects have this method. Non-string objects such as File, Integer, Float, `nil`, etc. all respond to `to_s`.

jeremywadsack replied:
Understood. This was the easier approach; I didn't know which direction you wanted to go. To me the "ruby" approach would be that anything that "acts like a" gs URL string is acceptable. 

`File`, `Integer`, `Float`, and `nil` would all fail the second half of the check:

```
file.to_s.downcase.start_with?("gs://"))
```

Happy to make the change for more explicit `URI`  support, just suggesting a less constrictive option.

blowmage replied:
That's a fair point. I'll leave it to @quartzmo to decide.

quartzmo replied:
I'm OK with any approach that works. More explicit `URI` support seems fine.

jeremywadsack replied:
The current approach works. Would you like me to change it to be more explicit/restrictive about `URI`? You're call.

quartzmo replied:
Yes, if you don't mind, let's go with the explicit URI support ("a conditional to the beginning of the method that checks if the object is a URI and convert it to a String." as @blowmage suggested). Thanks @jeremywadsack, appreciate your time and work on this!

blowmage commented on google-cloud-bigquery/lib/google/cloud/bigquery/table.rb:
```diff
@@ -1319,9 +1319,9 @@ def extract extract_url, format: nil, compression: nil, delimiter: nil,
-        # @param [File, Google::Cloud::Storage::File, String] file A file or the
-        #   URI of a Google Cloud Storage file containing data to load into the
-        #   table.
+        # @param [File, Google::Cloud::Storage::File, String, URI] file A file
```
To explicitly support an URI object, I would add a conditional to the beginning of the method that checks if the object is a URI and convert it to a String. Or, try to parse any string input with URI and handle the argument as a URI internally.

jeremywadsack commented:
@quartzmo @blowmage I've fixed this to only allow an explicit `URI` object and updated `Table#load`, `Dataset#load` and `Dataset#load_job` to accept that as well.

If that looks good I can squash and rebase.

quartzmo commented:
@jeremywadsack Can you rebase on master now for a fix to the CI build that I added yesterday (#1917)?
